app/api/utility.py

```diff
@@ -21,47 +18,76 @@ def add_trailing_slash(url: str) -> str:
     return url
 
 
-def parse_nodes_as_dict(nodes: str) -> list:
+def parse_nodes_as_dict(path: Path) -> dict:
     """
-    Transforms a string of user-defined Neurobagel nodes (from an environment variable) to a dict where the keys are the node URLs, and the values are the node names.
-    It uses a regular expression to match the url, name pairs.
+    Reads names and URLs of user-defined Neurobagel nodes from a JSON file (if available) and stores them in a dict
+    where the keys are the node URLs, and the values are the node names.
     Makes sure node URLs end with a slash.
     """
-    pattern = re.compile(r"\((?P<url>https?://[^)]+),\s?(?P<label>[^\)]+)\)")
-    matches = pattern.findall(nodes)
-    nodes_dict = {add_trailing_slash(url): label for url, label in matches}
-    return nodes_dict
+    # TODO: Add more validation of input JSON, including for JSONDecodeError (invalid JSON)
+    if path.exists() and path.stat().st_size > 0:
+        with open(path, "r") as f:
+            local_nodes = json.load(f)
+        if local_nodes:
+            if isinstance(local_nodes, list):
+                return {
+                    add_trailing_slash(node["ApiURL"]): node["NodeName"]
+                    for node in local_nodes
+                }
+            return {
+                add_trailing_slash(local_nodes["ApiURL"]): local_nodes[
+                    "NodeName"
+                ]
+            }
+
+    return {}
 
 
 async def create_federation_node_index():
     """
     Creates an index of nodes for federation, which is a dict where the keys are the node URLs, and the values are the node names.
     Fetches the names and URLs of public Neurobagel nodes from a remote directory file, and combines them with the user-defined local nodes.
     """
-    local_nodes = parse_nodes_as_dict(LOCAL_NODES)
-
     node_directory_url = "https://raw.githubusercontent.com/neurobagel/menu/main/node_directory/neurobagel_public_nodes.json"
+    local_nodes = parse_nodes_as_dict(LOCAL_NODE_INDEX_PATH)
+
+    if not local_nodes:
+        warnings.warn(
+            f"No local Neurobagel nodes found. Federation will be limited to nodes available from the Neurobagel public node directory {node_directory_url}. "
+            "(To specify one or more local nodes to federate over, define them in a 'local_nb_nodes.json' file in the current directory and relaunch the API.)\n"
+        )
```

Comment:
I think this warning is more `parse_nodes_as_dict`'s responsibility, no?

Reply:
This would also help make the warning less generic / confusing (are there no local nodes, did the file not exist, was the file empty ...)

Reply:
Since this change would require updating several tests, will address separately as part of https://github.com/neurobagel/federation-api/issues/41